File: zig/ui/glfw/vertex_array.zig
const std = @import("std");
const gl = @cImport(@cInclude("GL/glew.h"));
const vertex_buffer = @import("vertex_buffer.zig");
const vertex_buffer_layout = @import("vertex_buffer_layout.zig");
const VertexBuffer = vertex_buffer.VertexBuffer;
const VertexBufferLayout = vertex_buffer_layout.VertexBufferLayout;


pub const VertexArray = struct {
    rendererID: u32 = 0,



    pub fn destroy(self: *VertexArray) void {
        gl.glDeleteVertexArrays().?(1, &self.rendererID);
    }

    pub fn addBuffer(self: *VertexArray, vb: *VertexBuffer, layout: *VertexBufferLayout) void {
        vb.bind();
        self.bind();

        var offset: usize = 0;
        for(0.., layout.elements.items) |i, *element| {
            gl.glEnableVertexAttribArray().?(@intCast(i));
            gl.glVertexAttribPointer().?(@intCast(i), @intCast(element.count), element._type, @as(u8, @intFromBool(element.normalized)), 
                @intCast(layout.stride), @ptrFromInt(offset));
            offset += element.count * element.bufferElementGetSize();
        }
    }

    pub fn bind(self: *VertexArray) void {
        gl.glBindVertexArray().?(self.rendererID);

    }

    pub fn unbind(self: *VertexArray) void {
        _ = &self;
        gl.glBindVertexArray().?(0);
    }

};

pub fn vertexArray() VertexArray {
    var va: VertexArray = .{};
    gl.glGenVertexArrays().?(1, &va.rendererID);
    return va;
}

File: c++/ui/opengl/vertex_array.cpp
#include "vertex_array.hpp"
#include <GL/glew.h>


VertexArray::VertexArray() {
    glGenVertexArrays(1, &this->rendererID);
}

VertexArray::~VertexArray() {
    glDeleteVertexArrays(1, &this->rendererID);
}

void VertexArray::addBuffer(VertexBuffer &vb, VertexBufferLayout &vbl) {
    this->bind();
    vb.bind();

    long offset = 0;
    int i = 0;

    for(auto e = vbl.elements.begin(); e != vbl.elements.end(); e++, i++) {
        glEnableVertexAttribArray(i);
        glVertexAttribPointer(i, e->count, e->_type, e->normalized, vbl.stride, (void *)offset);
        offset += e->count * e->bufferElementGetSize();
    }
}

void VertexArray::bind() {
    glBindVertexArray(this->rendererID);
}

void VertexArray::unbind() {
    glBindVertexArray(0);
}
